### ISSUE TITLE
CMake: Every possible method to update c_cpp_properties is broken, and all existing Extensions on the marketplace are broken

### ISSUE THREAD

**Issue by Yzrsah** [microsoft/vscode-cpptools#2726]
Microsoft owns GitHub and CMake is the most prevalent build system for C++ libraries on GitHub. And yet *there is no possible way* to get your c_cpp_properties.json updated from CMake includes. Try it out for yourself and see.

Try searching for CMake extensions. Every extension which interacts with c_cpp_properties.json is broken.

I have just watched a good talk from Rong Lu about new C++ features in Visual Studio Code, but this is truly glossing over the details and lacks the perspective on what a developer would actually go through attempting to use VSCode for any real C++ development.

CMake Tools: Not-working with compile_commands
CMake Tools Helper: Broken with failing commands
cmake-clangd-helper: Non-functional and broken
nil_cpp_config_updater: Barely functional and broken
vscode-clangd: Lacks-functionality
clang-include-fixer: Non-functional

We need GitHub to be working. It doesn't work because includes always fail.

**Comment by sean-mcmanus** [comment]
I believe the CMake Tools extension works.

Another option is to use cmake and a generator that can output a compile_commands.json (and setting the compileCommands setting in your config in c_cpp_properties.json).

**Comment by Yzrsah** [comment]
CMake Tools extension is also broken.

If you use:
`"compileCommands": "${workspaceFolder}/build/compile_commands.json"`

VSCode will still outline everything in red or use the tag parser. Issue #2098

- Even if that did work, you have to *hardcode* in a ${workspaceFolder}/build/compile_commands.json location. This is not generic for GitHub projects and it has to be constantly edited or generated.
- When you use the command "C/Cpp: Edit Configurations..." to generate a new file, all values are wrong. cppStandard is wrong, compilerPath is wrong. Nothing from CMake is utilized.
- Broken extensions such as CMake Tools Helper (which attempted to fix CMake Tools extension) overwrite the entire c_cpp_properties file, without any awareness of compile_commands
- Even when everything is manually added to c_cpp_properties, it still fails at a basic level with semantic red-highlight warnings: "use of multiple declarators in a single using declaration is a C++17 extension" after we have declared "cppStandard": "c++17"

**Comment by Zyrsha** [comment]
I have a team of 5 developers here. I told them: "Go out and configure your VSCode for C++, try everything out and see what you can get working"

All 5 team members came back with failing includes and a broken setup. They told me "we need time to write new extensions"

From my observation, I would guess this is a recurring problem plaguing vscode-cpptools.

**Comment by Yzrsah** [comment]
I found the "C_Cpp.loggingLevel": "6" and it is helping greatly.
Now I can see some of the issues that are occurring.

The first issue was that compiler detection was incorrect!

If I use:
`"C_Cpp.default.compilerPath": "/usr/local/bin/clang++",`
...**Bug**: causes it to detect defaults from /usr/bin/clang instead!

But detection is now correct if I use:
`"C_Cpp.default.compilerPath": "/usr/local/opt/llvm/bin/clang++",`

They are the same file and /usr/local/bin/clang++ is a symlink to /usr/local/opt/llvm/bin/clang++

Another issue was that extensions can silently disable compile_commands usage. I.e. with "vscode-clangd" extension installed. It would be helpful if there was an error or notification when "compileCommands" is provided by the user but being ignored due to a LSP extension.

**Comment by bobbrow** [comment]
Could you elaborate on what is broken with the CMake Tools extension?  The recommendation is to enable CMake Tools' configuration provider to provide IntelliSense configurations instead of using compile_commands.json.

You say you are working on a GitHub project.  Is it public?  Can we clone it to debug your problem?

**Comment by RobRuckOB** [comment]
Just to give my contribution, I too have been on a similar quest for intellisense that works (Visual Studio has spoiled so many of us) and have found similar issues as @Zyrsha with C++ intellisense - Tag Parser works, kinda, but the Default engine just will not behave. @bobbrow, after seeing some of your suggestions above I've gone back and taken another look at it.

For me there are two overarching issues:

- Some header files will show up problems, presumably because when taken in isolation they may not contain explicit #include references to the things that they declare for use, i.e. a header file makes use of `std::vector`, but does not contain `#include <vector>`, because it's already been done elsewhere. If that presumption is correct, this can be resolved by each and every header file containing such includes, but it'd just be to support the intellisense system. The annoying part of the issue is that when looking at one of these problem headers, cmake-tools does seem to have provided include paths that do contain where to find things, but still problems come up.

- More worryingly, after clicking around a few .cpp and .h files I'll normally start to receive these kinds of errors in the output log:
```
IntelliSense client is currently disconnected
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info
Failed to read response from server: 20
IntelliSense engine is not responding. Using the Tag Parser instead.
```
Which is clearly a problem, and may well be the root of most of my problems. When these errors do start happening, Tag Parser doesn't seem to be entirely fallen back to and the whole situation seems to end up in a twilight between the two engines where neither one works quite right.

FWIW, my issues are in a sizeable code base that is unfortunately not public, and are present with these versions:

VS Code 1.28.2
Cpp tools 0.20.1
CMake tools 1.1.2

**Comment by sean-mcmanus** [comment]
@RobRuckOB "Failed to read response from server" means our IntelliSense process crashed, usually it's due to our parser. If you can attach a debugger to the Microsoft.VSCode.CPP.IntelliSense.msvc process to get a call stack of the crash and/or an isolated code snippet that causes the crash then that would help to fix it (on Mac, a crash log file is usually created by the OS). Our pending 0.21.0-insiders will have an updated parser which will have some crashes fixed, but we don't know yet if there will be regressions.

**Comment by bobbrow** [comment]
@RobRuckOB headers are a particularly difficult problem.  Before starting IntelliSense on a given header, our extension attempts to find an existing IntelliSense process that already includes that header or some  cpp file in your workspace that includes it so that the full context can be available.  If you are having trouble with IntelliSense on headers, you can open any cpp file that directly/indirectly includes the header in question and that should fix your header's IntelliSense.

**Comment by RobRuckOB** [comment]
@sean-mcmanus, I'd assumed that's what was happening, but thanks for confirming. I've attached a handful of Mac crash logs (.txt added to keep GH happy). And the relevant bit of assembly from lldb from when the crash occurs:

```
* thread #12, stop reason = EXC_BAD_ACCESS (code=1, address=0x24)
    frame #0: 0x0000000102f5b7ab Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin`check_defaulted_or_deleted_function + 427
Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin`check_defaulted_or_deleted_function:
->  0x102f5b7ab <+427>: cmpl   $0x2, 0x24(%rax)
    0x102f5b7af <+431>: sete   %bl
    0x102f5b7b2 <+434>: callq  0x1030823a0               ; void_type
    0x102f5b7b7 <+439>: movq   %rax, %rdi
```

[Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin_2018-11-06-143155_Robs-MacBook-Pro.crash.txt](https://github.com/Microsoft/vscode-cpptools/files/2556744/Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin_2018-11-06-143155_Robs-MacBook-Pro.crash.txt)
[Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin_2018-11-06-144619_Robs-MacBook-Pro.crash.txt](https://github.com/Microsoft/vscode-cpptools/files/2556745/Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin_2018-11-06-144619_Robs-MacBook-Pro.crash.txt)
[Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin_2018-11-06-143328_Robs-MacBook-Pro.crash.txt](https://github.com/Microsoft/vscode-cpptools/files/2556746/Microsoft.VSCode.CPP.IntelliSense.Msvc.darwin_2018-11-06-143328_Robs-MacBook-Pro.crash.txt)

@bobbrow, I can well imagine they're problematic. It sounds like if the intellisense process is going awry, when finding an existing process that includes the header, _might_ be exacerbating matters.

**Comment by sean-mcmanus** [comment]
@RobRuckOB It's possible our pending 0.21.0-insiders release may have a fix for for that crash, because it occurred with Visual Studio (on Windows) in September, but not after the October update.

**Comment by Zyrsha** [comment]
Hi all. I was never able to make VSCode work.
My C/C++ tools is still broken.

Here are the steps I used:

```
- install macOS Mojave

- install Xcode
- install Xcode command line developer tools
	xcode-select --install
	open /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg

- install Homebrew
	/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

- lldb_codesign certificate https://opensource.apple.com/source/lldb/lldb-69/docs/code-signing.txt
- brew install llvm --with-toolchain --with-lldb --with-python@2
- update .bash_profile
	export PATH="/usr/local/opt/llvm/bin:$PATH"
	export LDFLAGS="-L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib"
	export CPPFLAGS="-I/usr/local/opt/llvm/include"
	export LIBCLANG_SYSTEM_INCLUDE_DIR="/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include"

- verify include paths
	#include <iostream>
	#define SHOW(...) { std::cout << #__VA_ARGS__ " == " << __VA_ARGS__ << std::endl; }
	int main() { SHOW(__cplusplus); SHOW(__clang_version__); SHOW(_LIBCPP_VERSION); SHOW(_LIBCPP_STD_VER); }

- brew install cmake
- brew install ninja

- install Visual Studio Code
	ln -s "/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code" /usr/local/bin/code
- install VSCode extensions
	C/C++
	C/C++ Clang Command Adapter
	CMake
	CMake Tools
	ErrorLens
	ninja syntax

- configure CMake extension
	"cmake.cmakePath": "/usr/local/bin/cmake",
	"cmake.configureArgs": [
		"-DCMAKE_EXPORT_COMPILE_COMMANDS=ON"
	],
- configure C_Cpp extension
	"C_Cpp.default.compilerPath": "/usr/local/opt/llvm/bin/clang++",
	"C_Cpp.default.intelliSenseMode": "clang-x64",
	"C_Cpp.default.cppStandard": "c++17",
	"C_Cpp.default.cStandard": "c11",
	"C_Cpp.intelliSenseEngineFallback": "Disabled",

- configure CMake project for Clang's libc++
	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -stdlib=libc++")
	set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -stdlib=libc++ -lc++abi")

- CMake: Edit user-local CMake kits
	{ "name": "Clang 7.0.0",
	  "compilers": {
	    "C": "/usr/local/opt/llvm/bin/clang",
	    "CXX": "/usr/local/opt/llvm/bin/clang++"
	  }
	}
- CMake: Select a Kit
	Clang 7.0.0 kit
- CMake: Configure
- C/Cpp: Edit Configurations...
	"compileCommands": "${workspaceFolder}/build/compile_commands.json"

- Audit configuration
	"C_Cpp.loggingLevel": "6", // ! Verify that your setup is working, read carefully
```

There's a number of issues going on here.
- The editor's error highlighting is filled with walls of errors. The "compiler" that is being used for VSCode's editor never matches the one that CMake is building with.
- The VSCode build window shows 0 warnings, the output is perfect and everything seems to be correct. But the VSCode editor window is complaining that c++17 members of std do not exist, that headers do not exist, etc.
- Checking in my C/Cpp log, I can see that the c++17 standard is being selected, that my exact clang is being utilized, paths and values are correct. Even though C/Cpp extension thinks it's in c++17 mode, the editor is still displaying errors such as "error: auto not allowed in template parameter until c++17" etc
- I attempted to alter all settings that could potentially affect this, i.e. setting Intellisense mode to MSVC instead of Clang-x64.
- There was a bug. When providing a path to /usr/local/bin/clang, the symlink is not detected correctly and vscode uses /usr/bin/clang instead. A bunch of values in the log were incorrect. I fixed that. But VSCode editor still has a parallel set of disagreeing errors that do not match up with my build.
- Some extensions will silently cause compile_commands to stop being utilized and there's initially no way to be sure of what is going on. These extensions can also cause language support to stop working. i.e. CMake Tools Helper can shut off compile_commands. From the user perspective when first setting this up, there was no feedback about anything. I had a brokeb setup for days due to "CMake Tools Helper"
- This is confusing because it seems like a recommend extension, but we are unsure which stuff is even working or what the recommend extensions are.
- There's a bunch of low quality extensions available that can waste time and actually provide no outcome.
- I have now disabled all extensions except the few required essentials, because they provide uncertainty about breaking the setup. We need more information.
- I had not yet found the C_cpp.loggingLevel 6. Without that information, the user is in the dark about what is going on with their compiler.

When helping a user with these types of issues, there should be a standard and easy procedure to check their compiler setup. This should allow us to see a compact output with all of the relevant information about the compiler setup and state used by C/Cpp tools.

It might look like this:
- C_Cpp: Show Compiler Setup

So the first thing you could ask the users would be:
- What does your "Show C/C++ Compiler Setup" page say? Can you paste it here?
- What does your "Show Language Server Status" page say? Can you paste it here?
- Are you using the Microsoft recommended C/C++ and CMake extension set?
- Have you read the Wiki page about configuring a C++ and CMake toolchain?

This would be the go-to page with all information about the current state of compiler support
	- compiler mode - compiler path - compiler version - cppstandard version - stdlib version - language server status - list of language server hooks from extensions - intellisense mode - intellisense status (what's really being used?) - list of intellisense hooks from extensions - compile_commands status (is it even being used?!) - list of defines (not the extended list) - list of include paths - list of browse paths - project directory - build directory
- ... any other state that could possibly be affected by extensions.
- This could be locked down to an API point so it's not possible for assertions made by the Compiler Setup page to be incorrect. I.e. this page will reflect the correct status of the system after being altered by extensions.

We want to provide a 1-click CMake setup for VSCode, if at all possible. I'm not having any issues with other IDE's. They were able to detect and use CMake and Clang without issue. This is very important for VSCode because CMake is huge. I really hope to use VSCode in the future.

**Comment by Colengms** [comment]
There hasn't been activity on this issue for quite a while.  We've made many improvements in this area. 
 If you are still seeing a problem, please open a new issue, and please include repro details based on a recent build of the extension.  :)  Thanks.
